### PR TITLE
updated CIB, CO, and LRG models

### PR DESCRIPTION
**Pull request by dongwooc**
CIB models pending re-fitting with 'field' component

### CODE REVIEW

File: src/cib.jl [hunk 1]
```diff
@@ -35,24 +33,22 @@ not typed are converted to type T. This model has the following parameters and d
...
-    Inu_norm     = 0.3180384
     min_redshift = 0.0
     max_redshift = 5.0
     min_mass     = 1e12
     box_size     = 40000
 
     # shang HOD
     shang_zplat  = 2.0
-    shang_Td     = 20.7
+    shang_Td     = 20.7 #23.0
```
Comment (xzackli) [inline src/cib.jl:43]: What are these comments?

File: src/cib.jl [hunk 2]
```diff
@@ -151,27 +171,69 @@ end
...
+
+"""
+Quiescent fraction recipe from UniverseMachine
+"""
+function fquench_UM(Mh::T,z::T,model::AbstractCIBModel) where T
+    a = one(T)/(one(T)+z);
+    M200kms = T(1.64e12)/((a/T(0.378))^T(-0.142)+(a/T(0.378))^T(-1.79)) # MSol
+    vMpeak = T(200)*(Mh/M200kms)^T(0.3) # km/s
+    Qmin = max(T(0),T(model.quench_Qmin0)+T(model.quench_Qmina)*(one(T)-a));
+    logVQ = T(model.quench_VQ0) - T(model.quench_VQa)*(T(1)-a) + T(model.quench_VQz)*z;
```
Comment (xzackli) [inline src/cib.jl:191]: You really only need to convert the type for string literals, but this way of writing the code does ensure you stay in type T operations. Alternatively you could just change the type at the end.

Comment (xzackli) [reply src/cib.jl:191]: (this is fine, just bemoaning the T everywhere)

File: src/co_broadband.jl [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+            quasiTcoJ_sat[i,J]/= xRJ^2*exp(xRJ)/(exp(xRJ)-1)^2
+        end
+        nuCI_sat[i] = 492.16f0/(1.0f0+sources.redshift_sat[i])
+        LCI_sat[i] = LcoJ_sat[i,1]*model.R_CO10_CI*exp((randn()-0.5*2.302585*model.scatterdex_CO_CI)*2.302585*model.scatterdex_CO_CI)
+        quasiTCI_sat[i] = LCI_sat[i]*(1.315415f0/4.0f0/pi/nuCI_sat[i]^2.0f0/sources.dist_sat[i]^2.0f0/(1.0f0+sources.redshift_sat[i])^2.0f0) # divide by dnu in GHz
+        xRJ = model.xRJ_GHz*nuCI_sat[i]
+        quasiTCI_sat[i]/= xRJ^2*exp(xRJ)/(exp(xRJ)-1)^2
+    end
+    return (
+        N_cen = sources.N_cen,
```
Comment (xzackli) [inline src/co_broadband.jl:102]: someday we ought to document what these mean but it is not today

File: src/lrg.jl [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+    l10Mcut = log10.(model.zheng_Mcut)
+
+    for i in 1:size(N_cen_i,1)
+        for ii in 1:size(N_cen_i,2)
+            N_cen_i[i,ii] = model.yuan_ic[ii]/T(2)*SpecialFunctions.erfc((l10Mcut[ii]-x_m[i]/2.30258509)/(T(1.41421356)*model.zheng_sigma[ii]))
+            N_cen_i[i,ii] = convert(T, max(0.0, N_cen_i[i,ii]))
+        end
+    end
+
+    return [LinearInterpolation(x_m, N_cen_i[:,ii]) for ii in 1:size(N_cen_i,2)]
```
Comment (xzackli) [inline src/lrg.jl:89]: reject broadcasting
embrace tradition (for loops)

File: src/lrg.jl [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+        N_sh_i[i], err = quadgk( t->integrand_m(t, x_m[i]),
+                log(model.shang_Msmin), x_m[i], rtol=1e-8)
+        N_sh_i[i] = convert(T, max(0.0, N_sh_i[i]))
+    end
+
+    return LinearInterpolation(x_m, N_sh_i)
+end
+
+"""
+Build a linear interpolator for drawing masses from the subhalo mass function.
```
Comment (xzackli) [inline src/lrg.jl:140]: what are these

**Review by xzackli — APPROVED** [review]
Feel free to merge this into main, as long as you file a number of issues on the repo so we remember to do it later.

- (on publish) write a tutorial for LRG/etc like for CIB and tSZ in the docs
- consider use of [ChangePrecision](https://stackoverflow.com/questions/68068823/how-to-change-default-float-to-float32-in-a-local-julia-environment) or the like someday
- write end-to-end regression tests based on [StableRNGs](https://github.com/JuliaRandom/StableRNGs.jl)

File: src/cib.jl (hunk 1 above)
Comment (dongwooc) [reply src/cib.jl:43]: I think I have an appreciation for how junk DNA arises

File: src/lrg.jl (hunk 5 above)
Comment (dongwooc) [reply src/lrg.jl]: purged

File: src/cib.jl (hunk 1 above)
Comment (dongwooc) [reply src/cib.jl]: purged

File: src/lrg.jl (hunk 4 above)
Comment (dongwooc) [reply src/lrg.jl:89]: ~~Colchester, CT~~ Julia: where tradition meets tomorrow

**Pull request merged**